treewide: improve use of STAGING_DIR_HOST

Ansuel opened the pull request:
This evolved in a much bigger project.

It was notice an inconsistency between what .mk define and some relocatable patch.
Relocatable patch derive the path from STAGING_DIR
.mk hardcode everything to staging_dir/host
other use STAGING_DIR_HOST that is always hardcoded to staging_dir/host

The intention of this big pr is to change everything to use STAGING_DIR_HOST and to change how STAGING_DIR_HOST is defined.

With this pr STAGING_DIR_HOST is now derived from STAGING_DIR.

This is needed as some downstream project provide STAGING_DIR with the make to use a different staging_dir than the default one. The original implementation is saved while fixing the inconsistency between very .mk and tools patches.

---

~~Instead of using STAGING_DIR and then go up one dir with '../' use
directly STAGING_DIR_HOST env variable. This should produce cleaner
symbolic links.~~

Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

Ansuel committed:
automake: use STAGING_DIR_HOST in relocatable patch

Instead of using STAGING_DIR and then go up one dir with '../' use
directly STAGING_DIR_HOST env variable. This should produce cleaner
symbolic links.

Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

Ansuel committed:
autoconf: use STAGING_DIR_HOST in relocatable patch

Instead of using STAGING_DIR and then go up one dir with '../' use
directly STAGING_DIR_HOST env variable. This should produce cleaner
symbolic links.

Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

Ansuel committed:
libtool: use STAGING_DIR_HOST in relocatable patch

Instead of using STAGING_DIR and then go up one dir with '../' use
directly STAGING_DIR_HOST env variable. This should produce cleaner
symbolic links.

Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

Ansuel commented:
@nbd168 @hauke @jow- wonder if you can take a look at this?

jow- commented:
This will likely break existing downstream procedures (e.g. manually invoking the toolchain after exporting a suitable `STAGING_DIR` env variable) - users would now require to set another, additional variable for things to work. 

NACK from me as the little cleanup gains do not seem worth it.

Ansuel commented:
@jow- btw i'm checking rules.mk and I still need to check my theory but...
From what i can see

```
PACKAGE_DIR:=$(BIN_DIR)/packages
PACKAGE_DIR_ALL:=$(TOPDIR)/staging_dir/packages/$(BOARD)
BUILD_DIR:=$(BUILD_DIR_BASE)/$(TARGET_DIR_NAME)
STAGING_DIR:=$(TOPDIR)/staging_dir/$(TARGET_DIR_NAME)
BUILD_DIR_TOOLCHAIN:=$(BUILD_DIR_BASE)/$(TOOLCHAIN_DIR_NAME)
TOOLCHAIN_DIR:=$(TOPDIR)/staging_dir/$(TOOLCHAIN_DIR_NAME)
STAMP_DIR:=$(BUILD_DIR)/stamp
STAMP_DIR_HOST=$(BUILD_DIR_HOST)/stamp
TARGET_ROOTFS_DIR?=$(if $(call qstrip,$(CONFIG_TARGET_ROOTFS_DIR)),$(call qstrip,$(CONFIG_TARGET_ROOTFS_DIR)),$(BUILD_DIR))
TARGET_DIR:=$(TARGET_ROOTFS_DIR)/root-$(BOARD)
STAGING_DIR_ROOT:=$(STAGING_DIR)/root-$(BOARD)
STAGING_DIR_IMAGE:=$(STAGING_DIR)/image
BUILD_LOG_DIR:=$(if $(call qstrip,$(CONFIG_BUILD_LOG_DIR)),$(call qstrip,$(CONFIG_BUILD_LOG_DIR)),$(TOPDIR)/logs)
PKG_INFO_DIR := $(STAGING_DIR)/pkginfo

BUILD_DIR_HOST:=$(if $(IS_PACKAGE_BUILD),$(BUILD_DIR_BASE)/hostpkg,$(BUILD_DIR_BASE)/host)
STAGING_DIR_HOST:=$(TOPDIR)/staging_dir/host
STAGING_DIR_HOSTPKG:=$(TOPDIR)/staging_dir/hostpkg
```

STAGING_DIR_HOST is not derived from STAGING_DIR so if not passed an an env variable it will always point to the hardcoded path of `$(TOPDIR)/staging_dir/host` and host dir have its own stamp dir to check if a package needs to be rebuilt or not...

Again I still have to test this and I'm probably wrong and missing something but from what I can see the use of an env variable is only related to toolchains stuff and host tools will still be built on the default location in our buildroot

Ansuel commented:
Yep I made a quick test with
`make V=s -j1 STAGING_DIR=/home/ansuel/openwrt-todel2/openwrt/stagista`

![image](https://user-images.githubusercontent.com/20289090/205316546-ccab4f92-35b1-47ce-9e05-ebc6711852f1.png)

![image](https://user-images.githubusercontent.com/20289090/205316508-c84b7a53-77c4-4a98-b582-bccc9ff228a3.png)

As you can see toolchain and host still use the default directory... And that will result in a wrong symbolic link I think...

Ansuel commented:
Well that was quick... automake fails to compile
```
Written by David J. MacKenzie and Akim Demaille.
configure:2942: $? = 0
configure:2950: result: yes
configure:2957: checking whether autoconf works
configure:2964: cd conftest && autoconf -o /dev/null conftest.ac
/home/ansuel/openwrt-todel2/openwrt/staging_dir/host/bin/autoconf: 509: exec: /home/ansuel/openwrt-todel2/openwrt/stagista/../host/bin/autom4te: not found
configure:2967: $? = 127
configure:2976: result: no
configure:2979: error: The installed version of autoconf does not work.
    Please check config.log for error messages before this one.
```
[config.log](https://github.com/openwrt/openwrt/files/10141992/config.log)

Ansuel commented:
I'm following this https://openwrt.org/docs/guide-developer/toolchain/crosscompile with using https://downloads.cdn.openwrt.org/snapshots/targets/ath79/generic/openwrt-toolchain-ath79-generic_gcc-11.3.0_musl.Linux-x86_64.tar.xz and CONFIG_AUTOREBUILD=n

Same error... So I think this is a fix more than a cleanup

jow- commented:
Then STAGING_DIR_HOST should be derived from STAGING_DIR instead of introducing another variable all over the place. It is cleaner to have a single, authoritative variable instead of having to deal with two different variables.

Ansuel commented:
But the problem is that with STAGING_DIR we refer to the TARGET staging_dir and host and hostpkg (and even toolchain) have their own staging directory.

I'm aware that many GPL suggest to declare a staging_dir but with the current rules.mk with STAGING_DIR we don't refer to the directory but to the target directory...  (i remember having something with some technicolor gpl source)

I need to check if we made some changes to how this variable was used but if it's not the case then toolchain, host and hostpkg tools never used that...

So we have 2 solution:
- change staging_dir value and point to a generic dir... (but this will result in even more downstream problems)
- continue with the current implementation and correctly use STAGING_DIR_HOST for host tools

Ansuel commented:
Even lede-17.01 have the same declaration for STAGING_DIR
```
PACKAGE_DIR:=$(BIN_DIR)/packages
BUILD_DIR:=$(BUILD_DIR_BASE)/$(TARGET_DIR_NAME)
STAGING_DIR:=$(TOPDIR)/staging_dir/$(TARGET_DIR_NAME)
BUILD_DIR_TOOLCHAIN:=$(BUILD_DIR_BASE)/$(TOOLCHAIN_DIR_NAME)
TOOLCHAIN_DIR:=$(TOPDIR)/staging_dir/$(TOOLCHAIN_DIR_NAME)
STAMP_DIR:=$(BUILD_DIR)/stamp
STAMP_DIR_HOST=$(BUILD_DIR_HOST)/stamp
TARGET_ROOTFS_DIR?=$(if $(call qstrip,$(CONFIG_TARGET_ROOTFS_DIR)),$(call qstrip,$(CONFIG_TARGET_ROOTFS_DIR)),$(BUILD_DIR))
TARGET_DIR:=$(TARGET_ROOTFS_DIR)/root-$(BOARD)
STAGING_DIR_ROOT:=$(STAGING_DIR)/root-$(BOARD)
BUILD_LOG_DIR:=$(TOPDIR)/logs
PKG_INFO_DIR := $(STAGING_DIR)/pkginfo

BUILD_DIR_HOST:=$(if $(IS_PACKAGE_BUILD),$(BUILD_DIR)/host,$(BUILD_DIR_BASE)/host)
STAGING_DIR_HOST:=$(TOPDIR)/staging_dir/host
STAGING_DIR_HOSTPKG:=$(STAGING_DIR)/host
```

jow- commented:
Yeah, I understand it. So even more reason to reject this change without some fixing around `STAGING_DIR_HOST`. You're replacing values that previously got inferred from the given env variable (`STAGING_DIR`) with ones referring to a different variable not influence by the env variable used so far.

Ansuel commented:
> The various changes to replace `${STAGING_DIR}/../host/...` with `${STAGING_DIR_HOST}/...` are fine if at the same time we rework `rules.mk` to declare `STAGING_DIR_HOST` relative to `STAGING_DIR` in case `STAGING_DIR` is predefined through the environment.

Ok I missed this piece... and I'm more than happy to follow that, it just looks cleaner to me.

So just to make sure to keep the same behaviour (it's hard to understand to me since this is very broken currently)

With:
- STAGING_DIR from command line declared as `STAGING_DIR=/home/ansuel/openwrt-todel2/openwrt/stagista`

We should go up one dir (`../`) and then reference `host`

So the STAGING_HOST_DIR should result in `/home/ansuel/openwrt-todel2/openwrt/host`

This is an example of a broken implementation but it's how the thing works currently with the relocatable patch.
The inconsistency is in the fact that
rules.mk:
- hardcode the path to `/home/ansuel/openwrt-todel2/openwrt/staging_dir/host`
relocatable_patch:
- use STAGING_DIR and append `../host`

But OK. It seems the only way to ""handle"" this is to fix the inconsistency (to not destroy downstream project) and ignore any downstream project (since they are broken anyway if they have a broken path structure)(host dir not in the upper path referenced by STAGING_DIR)

This was just to summarize everything... I think I will put all these info in the commit tweaking the STAGING_DIR_HOST... And I think I should also do the same treatment to STAGING_DIR_HOSTPKG

Ansuel committed:
tools/mpc: use STAGING_DIR_HOST instead of hardcoding default

Use STAGING_DIR_HOST to reference the staging dir for host tools instead
of hardcoding it to the default location.

Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

Ansuel committed:
toolchain/gcc: use STAGING_DIR_HOST instead of hardcoding default

Use STAGING_DIR_HOST to reference staging host directory instead of
hardcoding it to default path.

Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

Ansuel committed:
toolchain/gdb: use STAGING_DIR_HOST instead of hardcoding default

Use STAGING_DIR_HOST to reference staging host directory instead of
hardcoding it to default path.

Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

Ansuel committed:
treewide: derive host and hostpkg path from STAGING_DIR

STAGING_DIR may be provided from command line. We currently hardcoded
STAGING_DIR_HOST and STAGING_DIR_HOSTPKG to the default location but we
currently have some relocatable patch that derive the path from
STAGING_DIR.

Fix this and correctly derive STAGING_DIR_HOST and STAGING_DIR_HOSTPKG
from STAGING_DIR.

The intention is to fix inconsistency from the relocatable patch and the
use of STAGING_DIR_HOST that is always hardcoded.
This with a wrong configuration may end up in broken state with some
host tools expecing a PATH from STAGING_DIR and others using library
from the default staging_dir/host path.

To save downstream project the original implementation is saved while
fixing the inconsistency between patch and .mk.

Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

Ansuel commented:
@hnyman @mpratt14 Can you take a look at this if you have some suggestion? :D 

@jow- if you can check this, I implemented what you suggested. Wonder if you have suggestion in how I derived the path from STAGING_DIR

mpratt14 commented:
looks like a good idea

currently there is more usage of `$(realpath ...)` than usage of `$(abspath ...)`, but I don't know which is "better" if at all

I'll take a deeper look later

Ansuel commented:
@mpratt14 i guess perf? I will do some research on google about what is the main difference...

Ansuel commented:
Ok main difference is that realpath solve symbolic links while abspath doesn't... Should we solve symbolic links? I guess no

Ansuel commented:
@jow- ok for you ? @hauke ?
